controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +59,793 @@ ControllerManager::load_controller(
     [&](auto loader)
     {return loader->is_available(controller_type);});
 
-  std::shared_ptr<controller_interface::ControllerInterface> controller(nullptr);
+  controller_interface::ControllerInterfaceSharedPtr controller(nullptr);
   if (it != loaders_.cend()) {
     controller = (*it)->create(controller_type);
   } else {
     const std::string error_msg("Loader for controller '" + controller_name + "' not found\n");
     RCLCPP_ERROR(get_logger(), "%s", error_msg.c_str());
     throw std::runtime_error(error_msg);
   }
+  ControllerSpec controller_spec;
+  controller_spec.c = controller;
+  controller_spec.info.name = controller_name;
+  controller_spec.info.type = controller_type;
 
-  return add_controller_impl(controller, controller_name);
+  return add_controller_impl(controller_spec);
 }
 
-std::vector<std::shared_ptr<controller_interface::ControllerInterface>>
-ControllerManager::get_loaded_controllers() const
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::load_controller(
+  const std::string & controller_name)
 {
-  return loaded_controllers_;
+  const std::string param_name = controller_name + ".type";
+  std::string controller_type;
+  if (!has_parameter(param_name)) {
+    declare_parameter(param_name, rclcpp::ParameterValue());
+  }
+  if (!get_parameter(param_name, controller_type)) {
+    RCLCPP_ERROR(get_logger(), "'type' param not defined for %s", controller_name.c_str());
+    return nullptr;
+  }
+  return load_controller(controller_name, controller_type);
+}
+
+controller_interface::return_type ControllerManager::unload_controller(
+  const std::string & controller_name)
+{
+  // get reference to controller list
+  int free_controllers_list = (current_controllers_list_ + 1) % 2;
```

Comment:
This should be the same logic that was used on ROS1. 

The idea is that the update loop is using the list A, so you do all your modifications on the list B.

At the end of the function, you change `current_controllers_list_` so the index points to list B, on the next loop the B list will be used, and A is available to be cleared.

Reply:
Sure, I'll add it.

Reply:
> This should be the same logic that was used on ROS1.

OK, this could be. I still do not understand this `free_controllers_list` variable. I will contact you directly, maybe you could help me to grap it. 

You can mark this as resolved.

Reply:
I have refactored the whole thing here: https://github.com/ros-controls/ros2_control/pull/139/commits/4c6a609282f21e1fe8d33adce0dc523cb28eadf3

Hopefully it makes the original code easier to understand.